### PR TITLE
Cache utilities

### PR DESCRIPTION
**Pull request by leogrim**

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/common/service/ServiceClient.scala [hunk 1]
```diff
@@ -153,4 +153,10 @@ trait ServiceClient extends Logging {
     val futures = Random.shuffle(urls(call.url)).take(teegree).map(callUrl(call, _, body)) //need to shuffle
     Future.firstCompletedOf(futures)
   }
+
+  ServiceClient.register.add(this)
+}
```
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/service/ServiceClient.scala:158]: @ymatsuda @andrewconner @eishay 
In such a pattern, how can I ensure that another thread won't access the object in ServiceClient.register before it's done initializing? I have tried using DelayedUnit but it messes up Guice's injector and breaks most of our tests.
It looks like a classic problem, is there a classic solution?

Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/service/ServiceClient.scala:158]: Try synchronized (will send docs when I get home)

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/service/ServiceClient.scala:158]: Synchronized on 'this'? The Register itself is synchronized for concurrency
purpose, but not sure how to synchronize the constructor

Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/service/ServiceClient.scala:158]: Oh, I probably misunderstood then. Are you wanting to block access to ServiceClient.register entirely until they're all registered, or just _while_ one is still being registered? Or something else?

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/service/ServiceClient.scala:158]: When the line ServiceClient.register.add(this) is executed, the object is not fully initialized yet (inheriting traits / class still have to be initialized). So I'm wondering what happens if someone tries to access the object through the register in this interval.

Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/service/ServiceClient.scala:158]: For anyone reading, Scala singleton constructors execute in a `static {  }` block, which does give thread-safe guarantees. So unless Guice is doing something dangerous with their classloader, this code is safe.

**Pull request merged**